contrib/create-waku/src/cli.ts

```diff
@@ -0,0 +1,171 @@
+  return /^(?:@[a-z0-9-*~][a-z0-9-*._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/.test(
+    projectName
+  );
+}
+
+function toValidPackageName(projectName: string) {
+  return projectName
+    .trim()
+    .toLowerCase()
+    .replace(/\s+/g, "-")
+    .replace(/^[._]/, "")
+    .replace(/[^a-z0-9-~]+/g, "-");
+}
+
+// if the dir is empty or not exist
+function canSafelyOverwrite(dir: string) {
+  return !fs.existsSync(dir) || fs.readdirSync(dir).length === 0;
+}
+
+async function init() {
+  const cwd = process.cwd();
+
+  let targetDir = "";
+  const defaultProjectName = "waku-project";
+
+  const CHOICES = fs.readdirSync("template");
+  let result: {
+    packageName: string;
+    shouldOverwrite: string;
+    chooseProject: string;
+  };
+
+  try {
+    result = await prompts(
+      [
+        {
+          name: "projectName",
+          type: "text",
+          message: "Project Name",
+          initial: defaultProjectName,
+          onState: (state: any) =>
+            (targetDir = String(state.value).trim() || defaultProjectName),
+        },
+        {
+          name: "shouldOverwrite",
+          type: () => (canSafelyOverwrite(targetDir) ? null : "confirm"),
+          message: `${targetDir} is not empty. Remove existing files and continue?`,
+        },
+        {
+          name: "overwriteChecker",
+          type: (values: any) => {
+            if (values === false) {
+              throw new Error(red("✖") + " Operation cancelled");
+            }
+            return null;
+          },
+        },
+        {
+          name: "packageName",
+          type: () => (isValidPackageName(targetDir) ? null : "text"),
+          message: "Package name",
+          initial: () => toValidPackageName(targetDir),
+          validate: (dir: string) =>
+            isValidPackageName(dir) || "Invalid package.json name",
+        },
+        {
+          name: "chooseProject",
+          type: "select",
+          message: "Choose a starter template",
+          choices: [
+            { title: "basic-template", value: CHOICES[0] },
+            { title: "async-template", value: CHOICES[1] },
+            { title: "promise-template", value: CHOICES[2] },
+          ],
+        },
+      ],
+      {
+        onCancel: () => {
+          throw new Error(red("✖") + " Operation cancelled");
+        },
+      }
+    );
+  } catch (cancelled) {
+    if (cancelled instanceof Error) {
+      console.log(cancelled.message);
+    }
+    process.exit(1);
+  }
+
+  const { packageName, shouldOverwrite, chooseProject } = result;
+
+  const root = path.resolve(targetDir);
+
+  if (shouldOverwrite) {
+    fse.emptyDirSync(root);
+  } else if (!fs.existsSync(root)) {
+    fs.mkdirSync(root);
+  }
+
+  const pkg = {
+    name: packageName ?? toValidPackageName(targetDir),
+    version: "0.0.0",
+  };
+
+  console.log("Setting up project...");
+  const templateRoot = path.join(__dirname, "../template");
+
+  const render = function render(templateName: string) {
```

Comment:
Can we avoid the function? I don't necessarily prefer a long function (`init` in this case), and we generally should split into smaller functions (better for testing), but this doesn't seem necessary and `render` sounds confusing (we just copy and patch json).

Reply:
Removed `render` as it is not necessary, splitting `init` in the same file or creating another file for the functions.